src/main/scala/scalaz/ziokeeper/VectorClock.scala

```diff
@@ -0,0 +1,63 @@
+package scalaz.ziokeeper
+
+import scala.annotation.tailrec
+import VectorClock._
+case class VectorClock[A](value: Map[A, Long]) extends AnyVal {
+  def time(mid: A): Long = value.getOrElse(mid, 0L)
+
+  def increment(mid: A, value: Long): VectorClock[A] =
+    VectorClock(this.value.updated(mid, time(mid) + value))
+
+  def compare(right: VectorClock[A]): CompareVectorClock = {
+    val leftWithDefault  = this.value.withDefaultValue(0L)
+    val rightWithDefault = right.value.withDefaultValue(0L)
+
+    @tailrec
+    def loop(keys: List[A], leftBigger: Boolean, rightBigger: Boolean): (Boolean, Boolean) =
+      keys match {
+        case _ if leftBigger && rightBigger => (leftBigger, rightBigger)
+        case Nil                            => (leftBigger, rightBigger)
+        case h :: t =>
+          val diff = leftWithDefault(h) - rightWithDefault(h)
+          loop(t, leftBigger || diff > 0, rightBigger || diff < 0)
+      }
+
+    val keys = leftWithDefault.keySet.union(rightWithDefault.keySet)
+
+    loop(keys.toList, leftBigger = false, rightBigger = false) match {
```

Comment:
This looks potentially non-deterministic (due to the use of `Set`), which will not satisfy laws for equality. 

Consider sorting by `A`, for which you'll need an ordering on `A`. This will be expensive and speaks to the fact that we ultimately need a different representation for `VectorClock`, namely, two arrays (one for `A`, sorted in order, and one for `Long`, with the same ordering). But for now the inefficiency can be ignored.

Reply:
you are right this is not most efficient implementation but I try to make this correct in the first place then optimize.